### PR TITLE
fix: race condition (#50)

### PR DESCRIPTION
**Pull request by d-kuro**
refs: https://github.com/argoproj/gitops-engine/issues/50

If you import the code from https://github.com/argoproj/gitops-engine/pull/52, the race condition will be gone.

```shell
# refs: https://github.com/argoproj/gitops-engine/pull/52
$ git status
On branch feature/fix-race-condition
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   pkg/sync/sync_context_test.go
        modified:   pkg/utils/kube/kubetest/mock.go

no changes added to commit (use "git add" and/or "git commit -a")

$ go test ./... --race
?       github.com/argoproj/gitops-engine       [no test files]
?       github.com/argoproj/gitops-engine/agent [no test files]
ok      github.com/argoproj/gitops-engine/pkg/cache     (cached)
?       github.com/argoproj/gitops-engine/pkg/cache/mocks       [no test files]
ok      github.com/argoproj/gitops-engine/pkg/diff      (cached)
?       github.com/argoproj/gitops-engine/pkg/engine    [no test files]
ok      github.com/argoproj/gitops-engine/pkg/health    (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync      (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync/common       (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync/hook (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync/hook/helm    (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync/ignore       (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync/resource     (cached)
ok      github.com/argoproj/gitops-engine/pkg/sync/syncwaves    (cached)
?       github.com/argoproj/gitops-engine/pkg/utils/errors      [no test files]
ok      github.com/argoproj/gitops-engine/pkg/utils/exec        (cached)
?       github.com/argoproj/gitops-engine/pkg/utils/io  [no test files]
?       github.com/argoproj/gitops-engine/pkg/utils/json        [no test files]
ok      github.com/argoproj/gitops-engine/pkg/utils/kube        (cached)
?       github.com/argoproj/gitops-engine/pkg/utils/kube/kubetest       [no test files]
?       github.com/argoproj/gitops-engine/pkg/utils/testing     [no test files]
?       github.com/argoproj/gitops-engine/pkg/utils/text        [no test files]
ok      github.com/argoproj/gitops-engine/pkg/utils/tracing     (cached)
```

Among many functions, using `sync.Mutex` to secure a lock was complex and difficult.
I took the approach of using `sync.Map`.
I don't think this is the best response, but I think it's good that race conditions are gone.
If you have any comments, please feel free to do so!

### CODE REVIEW

**Comment by ash2k** [comment]
Hey, you beat me to it! I was also working on fixing these race conditions. I think your branch does more and does it better as I haven't tried to simplify things as much - I'm not that familiar with the codebase. You can see what I've done here if you are interested https://github.com/argoproj/gitops-engine/compare/master...ash2k:ash2k/fixes

File: pkg/sync/sync_context.go [hunk 1]
```diff
@@ -753,60 +759,71 @@ func (sc *syncContext) runTasks(tasks syncTasks, dryRun bool) runState {
...
+							runStateMutex.Lock()
 							runState = failed
+							runStateMutex.Unlock()
 							sc.setResourceResult(t, "", common.OperationError, fmt.Sprintf("failed to delete resource: %v", err))
 						}
 					} else {
 						// if there is anything that needs deleting, we are at best now in pending and
 						// want to return and wait for sync to be invoked again
+						runStateMutex.Lock()
 						runState = pending
```
Comment (ash2k) [inline pkg/sync/sync_context.go:771]: I think apart from the data race that you are fixing here, there is also a logic error. One goroutine may set `runState` to `failed` but then another one to `pending` or the other way around. This non-deterministic behavior is probably not right, or is it? Please see https://github.com/argoproj/gitops-engine/commit/1e0e79249b8b390d0736f753a7d2d490c0a90c41 for a different approach.

File: pkg/sync/sync_context.go [hunk 2]
```diff
@@ -651,7 +651,7 @@ func (sc *syncContext) Terminate() {
 				sc.setResourceResult(task, "", common.OperationFailed, fmt.Sprintf("Failed to delete: %v", err))
 				terminateSuccessful = false
 			} else {
-				sc.setResourceResult(task, "", common.OperationSucceeded, fmt.Sprintf("Deleted"))
```
Comment (alexmt) [inline pkg/sync/sync_context.go:654]: 👍

File: pkg/sync/sync_context.go [hunk 3]
```diff
@@ -721,14 +723,17 @@ func (sc *syncContext) runTasks(tasks syncTasks, dryRun bool) runState {
 	{
 		var wg sync.WaitGroup
 		for _, task := range pruneTasks {
+			task := task
```
Comment (alexmt) [inline pkg/sync/sync_context.go:726]: 👍

**Comment by d-kuro** [comment]
@ash2k @alexmt 
Thank you for your comment!
I'm going to work on the code fix that was pointed out to me over the weekend.

**Comment by ash2k** [comment]
@d-kuro 👍 

I suggest to always run `go test -race`. ArgoCD's tests too!

**Comment by alexmt** [comment]
v1.6 testing looks good so far. We just want to run it internally a few more days and create 1.6 GA release. As soon as 1.6 is out I will test this PR's performance.

File: pkg/sync/sync_context.go (hunk 3 above)
Comment (ash2k) [reply pkg/sync/sync_context.go:726]: This line is not needed as `task` is passed to the closure (line 737/742) and it uses `t` instead.

File: pkg/sync/sync_context.go [hunk 4]
```diff
@@ -743,6 +748,7 @@ func (sc *syncContext) runTasks(tasks syncTasks, dryRun bool) runState {
 	if runState == successful && createTasks.Any(func(t *syncTask) bool { return t.needsDeleting() }) {
 		var wg sync.WaitGroup
 		for _, task := range createTasks.Filter(func(t *syncTask) bool { return t.needsDeleting() }) {
+			task := task
```
Comment (ash2k) [inline pkg/sync/sync_context.go:751]: This line is not needed, same as above.

File: pkg/sync/sync_context.go [hunk 5]
```diff
@@ -707,10 +707,12 @@ func (sc *syncContext) runTasks(tasks syncTasks, dryRun bool) runState {
 	sc.log.WithFields(log.Fields{"numTasks": len(tasks), "dryRun": dryRun}).Debug("running tasks")
 
 	runState := successful
+	var runStateMutex sync.Mutex
 	var createTasks syncTasks
 	var pruneTasks syncTasks
 
 	for _, task := range tasks {
+		task := task
```
Comment (ash2k) [inline pkg/sync/sync_context.go:715]: I wonder why is this needed?

File: pkg/cache/cluster.go [hunk 6]
```diff
@@ -32,11 +31,24 @@ const (
 )
 
 type apiMeta struct {
+	lock            sync.Mutex
 	namespaced      bool
 	resourceVersion string
```
[the remark this answers is not in the document]
Comment (d-kuro) [reply pkg/cache/cluster.go]: Fixed
https://github.com/argoproj/gitops-engine/pull/51/commits/8c8713d18297a41d000a25f580d324d753b898b8

File: pkg/cache/cluster.go [hunk 7]
```diff
@@ -376,31 +515,23 @@ func (c *clusterCache) startMissingWatches() error {
 	return nil
 }
 
-func runSynced(lock sync.Locker, action func() error) error {
-	lock.Lock()
-	defer lock.Unlock()
+func runSynced(action func() error) error {
```
[the remark this answers is not in the document]
Comment (d-kuro) [reply pkg/cache/cluster.go]: Thanks!
https://github.com/argoproj/gitops-engine/pull/51/commits/29224aaf138d6775fb55a25ce8a4fbfb8527119a

File: pkg/sync/sync_context.go (hunk 1 above)
Comment (d-kuro) [reply pkg/sync/sync_context.go]: I cherry-picked from your commit.
Thank you!

https://github.com/argoproj/gitops-engine/pull/51/commits/6a762b3b226aa4aa8be9d643dd7538633d4581e5
https://github.com/argoproj/gitops-engine/pull/51/commits/c5d445f642a1283a09aa0c0056274e1e6fe17997

File: pkg/sync/sync_context.go (hunk 3 above)
Comment (d-kuro) [reply pkg/sync/sync_context.go]: I think that's been fixed with this commit.
https://github.com/argoproj/gitops-engine/ pull/51/commits/c5d445f642a1283a09aa0c 0056274e1e6fe17997

File: pkg/sync/sync_context.go (hunk 4 above)
Comment (d-kuro) [reply pkg/sync/sync_context.go]: I understand.
Thank you comments.

I think that's been fixed with this commit.
https://github.com/argoproj/gitops-engine/ pull/51/commits/c5d445f642a1283a09aa0c 0056274e1e6fe17997

File: pkg/sync/sync_context.go (hunk 5 above)
Comment (d-kuro) [reply pkg/sync/sync_context.go]: I think that's been fixed with this commit.
https://github.com/argoproj/gitops-engine/pull/51/commits/c5d445f642a1283a09aa0c0056274e1e6fe17997

File: pkg/cache/cluster.go [hunk 8]
```diff
@@ -62,10 +61,24 @@ func list(resClient dynamic.ResourceInterface, opts metav1.ListOptions) (*unstru
...
 }
 
+func (a *apiMeta) LoadResourceVersion() (string, error) {
+	v := a.resourceVersion.Load()
+	version, ok := v.(string)
+	if ok {
+		return version, nil
+	}
+
+	return "", fmt.Errorf("stored type is invalid in resourceVersion")
```
Comment (ash2k) [inline pkg/cache/cluster.go:75]: This cannot happen and if it does it should be ok to panic because... it should never happen?
I'd do this instead

```go
str, _ := a.resourceVersion.Load().(string)
return str, nil
```
This will not error out and it also preserves current semantics where the field is an empty string if it was never set. It's also nicer to not need to check the error in every place where this method is called.

File: pkg/cache/cluster.go [hunk 9]
```diff
@@ -141,19 +154,56 @@ func NewClusterCache(config *rest.Config, opts ...UpdateSettingsFunc) *clusterCa
...
 
+type Resources struct {
+	sync.Map
+}
+
+func (r *Resources) StoreResources(key kube.ResourceKey, value *Resource) {
+	r.Store(key, value)
+}
+
+func (r *Resources) LoadResources(key kube.ResourceKey) (*Resource, error) {
```
Comment (ash2k) [inline pkg/cache/cluster.go:165]: Same as above - panicking should be ok if something unexpected is in this private field. And the error is not checked in a few places below. Why return it then? :)

File: pkg/cache/cluster.go [hunk 10]
```diff
@@ -167,6 +217,67 @@ type clusterCache struct {
 	eventHandlers               map[uint64]OnEventHandler
 }
 
+func (c *clusterCache) StoreApisMeta(key schema.GroupKind, value *apiMeta) {
+	c.apisMeta.Store(key, value)
+}
+
+func (c *clusterCache) LoadApisMeta(key schema.GroupKind) (*apiMeta, error) {
```
Comment (ash2k) [inline pkg/cache/cluster.go:224]: Same as above - panicking should be ok if something unexpected is in this private field.

File: pkg/cache/cluster.go [hunk 11]
```diff
@@ -167,6 +217,67 @@ type clusterCache struct {
...
+		return true
+	})
+	return cnt
+}
+
+func (c *clusterCache) StoreNsIndex(key string, value *Resources) {
+	c.nsIndex.Store(key, value)
+}
+
+func (c *clusterCache) LoadNsIndex(key string) (*Resources, error) {
```
Comment (ash2k) [inline pkg/cache/cluster.go:259]: Same as above - panicking should be ok if something unexpected is in this private field.

File: pkg/cache/cluster.go [hunk 12]
```diff
@@ -223,9 +334,12 @@ func (c *clusterCache) GetAPIGroups() []metav1.APIGroup {
...
 
-func (c *clusterCache) replaceResourceCache(gk schema.GroupKind, resourceVersion string, objs []unstructured.Unstructured, ns string) {
-	info, ok := c.apisMeta[gk]
-	if ok {
+func (c *clusterCache) replaceResourceCache(gk schema.GroupKind, resourceVersion string, objs []unstructured.Unstructured, ns string) error {
+	info, err := c.LoadApisMeta(gk)
+	if err != nil {
+		return err
+	}
+	if info != nil {
```
Comment (ash2k) [inline pkg/cache/cluster.go:342]: This `if` can be inverted to avoid so much indentation. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

File: pkg/cache/cluster.go [hunk 13]
```diff
@@ -611,8 +760,6 @@ func (c *clusterCache) EnsureSynced() error {
 		return c.syncError
 	}
 
-	c.lock.Lock()
-	defer c.lock.Unlock()
 	// before doing any work, check once again now that we have the lock, to see if it got
 	// synced between the first check and now
```
Comment (ash2k) [inline pkg/cache/cluster.go:764]: This comment and code above can be deleted now.

File: pkg/cache/cluster.go [hunk 14]
```diff
@@ -626,31 +773,50 @@ func (c *clusterCache) EnsureSynced() error {
...
 // GetNamespaceTopLevelResources returns top level resources in the specified namespace
-func (c *clusterCache) GetNamespaceTopLevelResources(namespace string) map[kube.ResourceKey]*Resource {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-	resources := make(map[kube.ResourceKey]*Resource)
-	for _, res := range c.nsIndex[namespace] {
-		if len(res.OwnerRefs) == 0 {
-			resources[res.ResourceKey()] = res
+func (c *clusterCache) GetNamespaceTopLevelResources(namespace string) *Resources {
+	res, _ := c.LoadNsIndex(namespace)
```
Comment (ash2k) [inline pkg/cache/cluster.go:777]: No error check here too. That's why I'm suggesting to just let it panic if the field contents is corrupted. The next line will panic anyway.

**Comment by d-kuro** [comment]
@ash2k 

> I think it might be not the best idea to use sync.Map because it makes it harder to write code:

I understand.
I too think the `Range()` of the sync map is not good and should be improved.
Initially, we decided to use sync.Map because of the complexity of locking on the map, but the Now that we've cleaned up the code to access it, we can use the mutex and map access approach with I think we can switch it up.

I'm going to try to change it back from sync map to map once.
Thanks!

File: pkg/cache/cluster.go [hunk 15]
```diff
@@ -779,25 +969,35 @@ func (c *clusterCache) onNodeUpdated(oldRes *Resource, un *unstructured.Unstruct
...
-	existing, ok := c.resources[key]
-	if ok {
-		delete(c.resources, key)
-		ns, ok := c.nsIndex[key.Namespace]
-		if ok {
-			delete(ns, key)
-			if len(ns) == 0 {
-				delete(c.nsIndex, key.Namespace)
+func (c *clusterCache) onNodeRemoved(key kube.ResourceKey) error {
+	existing, err := c.resources.LoadResources(key)
```
Comment (alexmt) [inline pkg/cache/cluster.go:978]: I agree with @ash2k regarding `sync.Map` usage.

I think we still need to have a lock here to make sure `resources` map and namespace index are synchronized.

File: pkg/cache/cluster.go (hunk 11 above)
Comment (alexmt) [inline pkg/cache/cluster.go:259]: I think `LoadNsIndex` is a private method. Please rename to `loadNsIndex`.

File: pkg/cache/cluster.go [hunk 16]
```diff
@@ -167,6 +217,67 @@ type clusterCache struct {
...
+func (c *clusterCache) ApisMetaLength() int {
+	var cnt int
+	c.apisMeta.Range(func(key, value interface{}) bool {
+		cnt++
+		return true
+	})
+	return cnt
+}
+
+func (c *clusterCache) StoreNsIndex(key string, value *Resources) {
```
Comment (alexmt) [inline pkg/cache/cluster.go:255]: Same as `LoadNsIndex`. Please rename to `storeNsIndex`.

**Comment by alexmt** [comment]
Hello @d-kuro . I'm very interested in completing and merging this PR. Please, let me know if you have time to continue working on it. 
I can help. E.g. I can send PR to your fork that addresses `sync.Map` related comments. let me know

**Comment by d-kuro** [comment]
@alexmt 
Sorry for the late reply.

I'm going to try migrating from SyncMap to map, as I wrote in my comment below.
https://github.com/argoproj/gitops-engine/pull/51#issuecomment-646390088

I may ask you to follow up if you try and get stuck, but I'll try to do it myself for now.
Thanks for your concern!

**Comment by d-kuro** [comment]
@ash2k cc: @alexmt 

I'm trying to work on migration from `sync.Map` to `map`.
but it suffers from the race condition of map range access and writing.

> fatal error: concurrent map iteration and map write

Using RLock for map range access and Lock for write is simple but for example the following code.
https://github.com/argoproj/gitops-engine/blob/master/pkg/cache/cluster.go#L241-L249

Deadlocked this function, because we are deleting the map in range access that requires a read lock.
Is there a good way to write this without using `SyncMap`?

```go
// c.resources is "map[kube.ResourceKey]*Resource"

// Need RWMutex.RLock()
for key := range c.resources {
	if key.Kind != gk.Kind || key.Group != gk.Group || ns != "" && key.Namespace != ns {
		continue
	}

	if _, ok := objByKey[key]; !ok {
		// Need RWMutex.Lock()
		// "delete(c.resources, key)" in a function
		c.onNodeRemoved(key)
	}
}
```

**Comment by ash2k** [comment]
@d-kuro Perhaps `replaceResourceCache()` can be left as is but then `watchEvents()` which calls it can grab the write lock before calling it?

I'd like to suggest to split this PR as it's getting quite big. Perhaps make all the changes to `pkg/cache/cluster.go` in a separate PR?

I've also looked at the `clusterCache` more closely and I'm wondering why was it written using list+watch+manual cache (`resources` field) and not using [informers](https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go)? It might be a good idea to refactor `clusterCache` to use informers instead - they support notifications, indexes and are thoroughly tested by kubernetes inself. WDYT?

**Comment by alexmt** [comment]
Thanks a lot for continue working on it  @d-kuro !


I've looked only more time where `replaceResourceCache` is used. The method is executed in two places and both have write lock:

https://github.com/argoproj/gitops-engine/blob/ce9616ad10da6d05c78ecceba23ae256ccb0c0f2/pkg/cache/cluster.go#L443

https://github.com/argoproj/gitops-engine/blob/ce9616ad10da6d05c78ecceba23ae256ccb0c0f2/pkg/cache/cluster.go#L371


I suspect race condition happens in tests only: [TestNamespaceModeReplace](https://github.com/argoproj/gitops-engine/blob/7d3da9f16e353692824a6e83fc31de113b6a9251/pkg/cache/cluster_test.go#L378). I hope I'm not missing anything. If this is the case we just need to use lock in `TestNamespaceModeReplace` test.

**Comment by alexmt** [comment]
@ash2k we could not use informers framework because of informers cache the whole resource manifest. We need only resources reference and parent links. Switching to informers would increase memory consumption.

**Comment by ash2k** [comment]
@alexmt 

> we could not use informers framework because of informers cache the whole resource manifest.

This is the default behavior, but it is configurable. If only certain fields are required, a custom Go type can be crafted to use for any watched resource type. That way only the needed fields are unmarshaled and nothing else. For example, there is a whole package just for [working with metadata of objects](https://github.com/kubernetes/client-go/tree/master/metadata). Tuned client and informer are available there.

However, memory consumption will still be higher in the above case vs how things currently work because e.g. all annotations will be stored in the informer's cache. Thank you for answering the question :)

**Comment by alexmt** [comment]
Changes of this PR got split into several other PRs and race detection is now enabled in the project. @d-kuro , @ash2k thanks a lot for driving it!

**Comment by alswl** [comment]
`c.lock` and `c.listSemaphore` had dead lock situation in big clusters. `c.lock` holds too big scope in `sync()` function, we should remove `c.lock`, using `sync.Map`.
